text/3700-ptr-assume-moved.md

```diff
@@ -0,0 +1,257 @@
+aliasing rules - using two 'live' pointers which have different 64-bit addresses but do point to
+the same chunk of memory would weaken alias analysis and related optimisations.
+
+We can avoid this issue by assuming that a move from the untagged address to the tagged address has
+happened. To do so, we need the helper function to return a pointer that will be annotated in LLVM
+IR as [noalias](https://llvm.org/docs/LangRef.html#noalias), as per the following excerpt.
+
+> On function return values, the noalias attribute indicates that the function acts like a system
+> memory allocation function, returning a pointer to allocated storage disjoint from the storage for
+> any other object accessible to the caller.
+
+This will result in the new pointer getting a brand new provenance, disjoint from the provenance of
+the original pointer.
+
+Every change to the high bits has to at least simulate a move operation and we must ensure the old
+pointers are invalidated. This is due to the aforementioned discrepancy between how Rust & LLVM see
+a memory address and how the OS & hardware see memory addresses.  From the OS & hardware
+perspective, the high bits are reserved for metadata and do not actually form part of the address
+(in the sense of an 'address' being an index into the memory array).  From the LLVM perspective, the
+high bits are part of the address and changing them means we are now dealing with a different
+address altogether.  Having to reconcile those two views necessarily creates some friction and extra
+considerations.
+
+Function signature, documentation and implementation:
+
+```
+/// Assume that the object pointed to by a pointer has been moved to a new address
+///
+/// Intended for use with pointer tagging architecture features such as AArch64 TBI.
+/// This function creates a new pointer with the address `new_address` and a brand new provenance,
+/// then assumes that a move from the original address to the new address has taken place.
+/// Note that this is only an indication for the compiler - nothing actually gets moved or reallocated.
+///
+/// SAFETY: Users *must* ensure that `new_address` actually contains the same memory as the original.
+/// The primary use-case is working with various architecture pointer tagging schemes, where two
+/// different 64-bit addresses can point to the same chunk of memory due to some bits being ignored.
+/// When used incorrectly, this function can be used to violate the memory model in arbitrary ways.
+/// Furthermore, after using this function, users must ensure that the underlying memory is only ever
+/// accessed through the newly created pointer. Any accesses through the original pointer
+/// (or any pointers derived from it) would be Undefined Behaviour.
+/// Additionally, this function is only designed for use with heap allocations. Trying to use it with
+/// a pointer to a stack-allocated variable will almost certainly result in Undefined Behaviour and
```

Comment:
```suggestion
/// a pointer to a stack-allocated variable will result in Undefined Behaviour and
```
No need to perform hedging here.

The most obvious way this goes wrong is that when the stack frame is popped, we conceptually perform deallocation of the object using the old pointer, which as described above is UB.